File: cses/08_string_algorithms/2106_repeating-substring/solution.py
import sys
from itertools import accumulate


def solve() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        sys.stdout.write("-1\n")
        return

    s = data[0]                      # làm việc trực tiếp trên bytes cho nhanh
    n = len(s)
    if n <= 1:
        # Xâu rỗng hoặc một ký tự không thể có substring lặp lại
        sys.stdout.write("-1\n")
        return

    # --- Nén tọa độ các ký tự ban đầu về 0..d-1 (d = số ký tự phân biệt) ---
    order = sorted(set(s))
    cmap = {c: i for i, c in enumerate(order)}
    rank = [cmap[c] for c in s]

    # --- Xây dựng suffix array bằng prefix doubling ---
    # Mỗi bước gấp đôi độ dài tiền tố so sánh; khóa kết hợp gộp cặp (rank[i], rank[i+k])
    # thành một số nguyên: key[i] = rank[i]*base + (rank[i+k]+1) với base = n+1 > mọi giá trị phụ.
    base = n + 1
    sa = list(range(n))
    k = 1
    while True:
        rank_shift = rank[k:] + [-1] * k
        keys = [a * base + b + 1 for a, b in zip(rank, rank_shift)]
        sa = sorted(range(n), key=keys.__getitem__)

        # Gán lại rank mới: tăng 1 mỗi khi khóa đổi so với hậu tố liền trước trong sa
        sk = [keys[i] for i in sa]
        diff = [0] + [1 if sk[i] != sk[i - 1] else 0 for i in range(1, n)]
        rnk_sorted = list(accumulate(diff))       # rank mới theo thứ tự trong sa

        newrank = [0] * n
        for p, rv in zip(sa, rnk_sorted):
            newrank[p] = rv
        rank = newrank

        if rnk_sorted[-1] == n - 1:               # tất cả rank phân biệt -> sa đã đúng
            break
        k <<= 1

    # --- LCP array bằng thuật toán Kasai, tìm LCP kề nhau lớn nhất ---
    # Biến h giảm tối đa 1 mỗi bước nên tổng chi phí amortized O(n).
    lcp_max = 0
    lcp_pos = 0
    h = 0
    for i in range(n):
        r = rank[i]
        if r > 0:
            j = sa[r - 1]             # hậu tố kề trước i trong thứ tự từ điển
            while i + h < n and j + h < n and s[i + h] == s[j + h]:
                h += 1
            if h > lcp_max:
                lcp_max = h
                lcp_pos = i
            if h:
                h -= 1
        else:
            h = 0

    # max LCP = 0 nghĩa là không có substring nào lặp lại
    if lcp_max == 0:
        sys.stdout.write("-1\n")
    else:
        sys.stdout.buffer.write(s[lcp_pos:lcp_pos + lcp_max] + b"\n")


solve()

File: cses/08_string_algorithms/2106_repeating-substring/solution.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string text;
    if (!(std::cin >> text)) {
        return 0;
    }

    const int length = static_cast<int>(text.size());
    if (length == 1) {
        // Xâu một ký tự không thể có substring lặp lại
        std::cout << -1 << '\n';
        return 0;
    }

    // suffix_array: hoán vị các vị trí hậu tố; rank: xếp hạng ban đầu theo ký tự đầu
    std::vector<int> suffix_array(length);
    std::iota(suffix_array.begin(), suffix_array.end(), 0);
    std::vector<int> rank(length);
    for (int index = 0; index < length; ++index) {
        rank[index] = text[index] - 'a';
    }

    // Xây dựng suffix array bằng prefix doubling: mỗi bước gấp đôi độ dài tiền tố so sánh
    std::vector<int> next_rank(length);
    for (int step = 1;; step *= 2) {
        // So sánh hai hậu tố theo cặp (rank[i], rank[i+step]) tương ứng 2*step ký tự đầu
        const auto less_suffix = [&](const int left, const int right) {
            if (rank[left] != rank[right]) {
                return rank[left] < rank[right];
            }
            const int left_second = left + step < length ? rank[left + step] : -1;
            const int right_second = right + step < length ? rank[right + step] : -1;
            return left_second < right_second;
        };
        std::sort(suffix_array.begin(), suffix_array.end(), less_suffix);

        // Gán lại rank mới: tăng 1 mỗi khi cặp khóa đổi so với hậu tố liền trước trong SA
        next_rank[suffix_array[0]] = 0;
        for (int index = 1; index < length; ++index) {
            const int previous = suffix_array[index - 1];
            const int current = suffix_array[index];
            const int previous_second = previous + step < length ? rank[previous + step] : -1;
            const int current_second = current + step < length ? rank[current + step] : -1;
            const bool different = rank[previous] != rank[current] || previous_second != current_second;
            next_rank[current] = next_rank[previous] + (different ? 1 : 0);
        }
        rank.swap(next_rank);

        // Mọi rank đã phân biệt -> suffix array hoàn chỉnh
        if (rank[suffix_array.back()] == length - 1) {
            break;
        }
    }

    // LCP theo thuật toán Kasai: duyệt theo vị trí, biến common giảm tối đa 1 mỗi bước.
    // Với mỗi hậu tố tính LCP với hậu tố kề trước trong SA và giữ giá trị lớn nhất.
    int best_length = 0;
    int best_position = 0;
    int common = 0;
    for (int position = 0; position < length; ++position) {
        const int order = rank[position];
        if (order == 0) {
            common = 0;
            continue;
        }
        const int previous_suffix = suffix_array[order - 1];
        while (position + common < length && previous_suffix + common < length &&
               text[position + common] == text[previous_suffix + common]) {
            ++common;
        }
        if (common > best_length) {
            best_length = common;
            best_position = position;
        }
        if (common > 0) {
            --common;
        }
    }

    // max LCP = 0 nghĩa là không có substring nào lặp lại
    if (best_length == 0) {
        std::cout << -1 << '\n';
    } else {
        std::cout << text.substr(best_position, best_length) << '\n';
    }

    return 0;
}
